README.md
# README

## What is `xprintidle-ng`?

### [Web page][website]

[![Gitter][badge-gitter]][gitter]
[![Travis][badge-travis]][travis]
![Issues][badge-gh-issues]
![Forks][badge-gh-forks] 
![Stars][badge-gh-stars] 
![License][badge-license]

`xprintidle-ng` is a utility that queries the X server for the user's idle time
and prints it in milliseconds (or with a custom format string).

## Installation

Basically, type `./configure && make && sudo make install` to compile and
install the program. If you are compiling from the git repository, you must
run `./bootstrap` beforehand.

## Dependencies

### Build requirements

The following common executables must be available to build `xprintidle-ng`:
  * `cat`, `sed`, `tr`, `grep`, `mkdir`, `find`
  * `git`, `wget`, `perl`, `gzip`, `sha1sum`
  * `bash`, `gcc`, `make`

These executables are less common and are also required:
  * `autoconf`, `automake`, `autopoint`, `autoreconf`, `autom4te`, `libtoolize`
  * `pkg-config`, `m4`, `gettext` et al.
  * `help2man`, `makeinfo`

### Runtime requirements

The following libraries must be available to run `xprintidle-ng`:
  * `libxss` (also known as `libXScrnSaver` or similar)
  * `libXext`
  * `libX11`

Here is a short summary of packages required on various systems

| System     | Packages                                                        |
| ---------- | --------------------------------------------------------------- |
| Debian     | `libxss-dev`                                                    |
| Arch Linux | `libxss`                                                        |
| Gentoo     | `x11-base/xorg-server`                                          |
| NixOS      | `xlibs.libX11`, `xlibs.libXScrnSaver`, `xlibs.libXext`          |

## Contribute

### Translations

If you want to submit a translation for this project, here are the steps:
  1.  Install all relevant dependencies
  2.  Fork this repository on GitHub
  3.  `git clone http://github.com/<USERNAME>/xprintidle-ng && cd xprintidle-ng`
  4.  `./bootstrap && ./configure && make`
  5.  The template file will be in `po/xprintidle-ng.pot`
  6.  Based on the template file, write your `po` file, e.g.: `la.po` for Latin
  7.  Put `la.po` in the `po` folder in the repository
  8.  `git add po/la.po`
  9.  `git commit -a -m "Added Latin translation" && git push`
  10. Open a pull request against this repository

You can also translate `doc/xprintidle-ng.texi`; put these translations in an
appropriately named file, e.g.: `doc/xprintidle-ng.la.texi`.

Languages translated so far:
  * English




[website]:         https://taktoa.github.io/xprintidle-ng
[gitter]:          https://gitter.im/taktoa/xprintidle-ng
[travis]:          https://travis-ci.org/taktoa/xprintidle-ng

[badge-gitter]:    https://img.shields.io/badge/gitter-%E2%86%92-blue.svg
[badge-travis]:    https://img.shields.io/travis/taktoa/xprintidle-ng.svg
[badge-gh-issues]: https://img.shields.io/github/issues/taktoa/xprintidle-ng.svg
[badge-gh-forks]:  https://img.shields.io/github/forks/taktoa/xprintidle-ng.svg
[badge-gh-stars]:  https://img.shields.io/github/stars/taktoa/xprintidle-ng.svg
[badge-license]:   https://img.shields.io/badge/license-GPL-blue.svg


NEWS.md
xprintidle-ng 0.3:
* Switched to pkgconfig-based library detection
* Created texinfo/man documentation
* Added bootstrap script
* Added internationalization
* Added `--format` option

xprintidle 0.2:
* Now licensed under GNU GPL 2 (only).
* Workaround for incorrect idle times when changing DPMS state.

xprintidle 0.1:
* Initial version.


src/system.h
// system.h: system-dependent declarations; include this first.
// 
//   Copyright © 1996, 2005, 2006, 2007, 2008, 2013, 2014, 2015
//   Free Software Foundation, Inc.
// 
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation; either version 3, or (at your option)
//   any later version.
// 
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
//   GNU General Public License for more details.
// 
//   You should have received a copy of the GNU General Public License
//   along with this program. If not, see <http://www.gnu.org/licenses/>.

#ifndef XPRINTIDLE_NG_SYSTEM_H
#define XPRINTIDLE_NG_SYSTEM_H

// Assume ANSI C89 headers are available.
#include <limits.h>
#include <locale.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

// Use POSIX headers, substituting gnulib headers if they are not available
#include <getopt.h>
#include <unistd.h>

// Unicode
#include <wchar.h>
#include <wctype.h>

// Internationalization
#include "gettext.h"
#define _(str) gettext (str)
#define N_(str) gettext_noop (str)

// Check for errors on write.
#include "closeout.h"

// Just like strncmp, but the second argument must be a literal string
// and you don't specify the length; that comes from the literal.
#define STRNCMP_LIT(s, literal) \
  strncmp (s, "" literal "", sizeof (literal) - 1)

#endif /* XPRINTIDLE_NG_SYSTEM_H */


src/time-format.c
// Utilities for the rendering of elapsed time format strings
// 
// Copyright © 2015       Remy Goldschmidt        <taktoa@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of version 2 of the GNU General Public License
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to
//
//     Free Software Foundation, Inc.,
//     51 Franklin Street, Fifth Floor
//     Boston, MA 02110-1301, USA.

#include <stdbool.h>
#include <signal.h>
#include <time.h>
#include <config.h>
#include "system.h"
#include "xalloc.h"


// Is the current position in a given string escaped with
// an even number of escape characters?
static char is_escaped(const char escape_char, char* pos, char* beg) {
    char* i;
    for(i = pos - 1; i != beg && *i == escape_char; i--);
    return (pos - i) % 2;
}

// String trie
typedef struct trie_node_t trie_node_t;
struct trie_node_t {
    char* replace;
    trie_node_t* children[256];
}; 

// Struct representing a pair of strings
typedef struct str_pair_t str_pair_t;
struct str_pair_t {
    char* pattern;
    char* replacement;
}; 

// Allocate a trie and fill it with zeros
static trie_node_t* calloc_trie() {
    return calloc(1, sizeof(trie_node_t));
}

// Add a pattern-replacement pair to a trie
static void add_word_trie(trie_node_t* root,
                          char* pattern,
                          char* replacement) {
    if(!*pattern) {
        root->replace = replacement;
    } else {
        trie_node_t** children_arr = root->children;
        trie_node_t**  child = children_arr + *pattern;
        if(*child) {
            add_word_trie(*child, pattern + 1, replacement);
        } else {
            *child = calloc_trie();
            add_word_trie(root, pattern, replacement);
        }
    }    
}

// Is the given str_pair_t null?
static char not_null_pair(str_pair_t pair) {
    return (pair.pattern != NULL) && (pair.replacement != NULL);
}

// Generate a trie from an array of pairs of the given length
static trie_node_t* generate_trie(str_pair_t* pairs) {
    trie_node_t* root = calloc_trie();
    for(int i = 0; not_null_pair(pairs[i]); i++) {
        add_word_trie(root, pairs[i].pattern, pairs[i].replacement);
    }
    return root;
}

// Utility function for print_trie
static void indent_depth(int indent) {
    for(int i = 0; i < indent; i++) {
        fprintf(stdout, " ");
    }
}

// Print a trie, for debug purposes
static void print_trie(int indent, trie_node_t* trie) {
    if(trie->replace) {
        indent_depth(indent);
        fprintf(stdout, "Replacement: %s\n", trie->replace);
    }
    for(int i = 0; i < 256; i++) {
        if(trie->children[i]) {
            indent_depth(indent);
            fprintf(stdout, "Child: %i | %c\n", i, i);
            print_trie(indent + 1, trie->children[i]);
        }
    }
}

// Match a single character with a trie
static trie_node_t* trie_match(char character, trie_node_t* trie) {
    if(trie->children[character]) {
        return trie->children[character];
    } else {
        return trie;
    }
}

// Replace values in a string according to a trie
static char* trie_replace(char* input, trie_node_t* trie, char no_escape) {    
    for(int i = 0; i < strlen(input); i++) {
        if(trie_match(input[i], trie) != trie &&
           (no_escape || !is_escaped('\\', input + i, input))) {
            
            char* replace = NULL;
            int pat_len = -1;
            trie_node_t* node = trie;
            
            for(int j = i; j < strlen(input); j++) {
                trie_node_t* newnode = trie_match(input[j], node);
                if(node != newnode) {
                    node = newnode;
                    if(newnode->replace) {
                        replace = newnode->replace;
                        pat_len = j - i;
                        break;
                    }
                } else {
                    break;
                }
            }
            
            if(pat_len >= 0 && replace) {
                int str_len = strlen(input);
                int rep_len = strlen(replace);
                if((rep_len - pat_len) > 0) {
                    size_t new_len = 2 * (str_len + rep_len - pat_len);
                    input = realloc(input, sizeof(char) * new_len);
                }
                memmove(input + i + rep_len,
                        input + i + pat_len + 1,
                        str_len - i);
                memcpy(input + i, replace, rep_len);
                i += rep_len - 1;
            }
        }
    }
    size_t size = strlen(input);
    input = xrealloc(input, strlen(input));
    return input;
}

// Replace the pairs in the given input, optionally ignoring escaped characters
static char* string_replace(char* input, str_pair_t* pairs, char no_escape) {
    trie_node_t* trie = generate_trie(pairs);
    char* retval = trie_replace(input, trie, no_escape);
    free(trie);
    return retval;
}

// Format a time in milliseconds against a format string
char* timefmt(const char* format, ulong millis) {
    char* fmt = xstrdup(format);

    char str_rem_milliseconds_int[40];
    char str_rem_seconds_int[40];
    char str_rem_minutes_int[40];
    char str_rem_hours_int[40];
    char str_total_milliseconds[40];
    char str_total_seconds_int[40];
    char str_total_seconds_float[40];
    char str_total_minutes_int[40];
    char str_total_minutes_float[40];
    char str_total_hours_int[40];
    char str_total_hours_float[40];
    char str_total_days_int[40];
    char str_total_days_float[40];
    char str_start_time_millis[40];
    char str_start_time_seconds[40];

    ulong start_time_millis = (time(NULL) * 1000) - millis;
    
    sprintf(str_rem_milliseconds_int, "%03lu", (millis /       1) % 1000);
    sprintf(str_rem_seconds_int,      "%02lu", (millis /    1000) % 60);
    sprintf(str_rem_minutes_int,      "%02lu", (millis /   60000) % 60);
    sprintf(str_rem_hours_int,        "%02lu", (millis / 3600000) % 24);
    sprintf(str_total_milliseconds,   "%lu",   millis);
    sprintf(str_total_seconds_int,    "%lu",   millis / 1000);
    sprintf(str_total_seconds_float,  "%f",    millis / 1000.0);
    sprintf(str_total_minutes_int,    "%lu",   millis / 60000);
    sprintf(str_total_minutes_float,  "%f",    millis / 60000.0);
    sprintf(str_total_hours_int,      "%lu",   millis / 3600000);
    sprintf(str_total_hours_float,    "%f",    millis / 3600000.0);
    sprintf(str_total_days_int,       "%lu",   millis / 86400000);
    sprintf(str_total_days_float,     "%f",    millis / 86400000.0);
    sprintf(str_start_time_millis,    "%lu",   start_time_millis);
    sprintf(str_start_time_seconds,   "%lu",   start_time_millis / 1000);

    str_pair_t pairs[] = {
        (str_pair_t) { "%rms",  str_rem_milliseconds_int },
        (str_pair_t) { "%rsc",  str_rem_seconds_int },
        (str_pair_t) { "%rmn",  str_rem_minutes_int },
        (str_pair_t) { "%rhr",  str_rem_hours_int },
        (str_pair_t) { "%tms",  str_total_milliseconds },
        (str_pair_t) { "%tsci", str_total_seconds_int },
        (str_pair_t) { "%tscf", str_total_seconds_float },
        (str_pair_t) { "%tmni", str_total_minutes_int },
        (str_pair_t) { "%tmnf", str_total_minutes_float },
        (str_pair_t) { "%thri", str_total_hours_int },
        (str_pair_t) { "%thrf", str_total_hours_float },
        (str_pair_t) { "%tdyi", str_total_days_int },
        (str_pair_t) { "%tdyf", str_total_days_float },
        (str_pair_t) { "%stms", str_start_time_millis },
        (str_pair_t) { "%stsc", str_start_time_seconds },
        (str_pair_t) { NULL, NULL }
    };

    return string_replace(fmt, pairs, 1);
}


src/time-format.h
// Headers for time-format.c
// 
// Copyright © 2015       Remy Goldschmidt        <taktoa@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of version 2 of the GNU General Public License
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to
//
//     Free Software Foundation, Inc.,
//     51 Franklin Street, Fifth Floor
//     Boston, MA 02110-1301, USA.

#ifndef XPRINTIDLE_NG_TIME_FORMAT_H
#define XPRINTIDLE_NG_TIME_FORMAT_H

char* timefmt(const char* format, unsigned long milliseconds);

#endif /* XPRINTIDLE_NG_TIME_FORMAT_H */


src/xprintidle-ng.c
// This program prints the "idle time" of the user to stdout.  The "idle
// time" is the number of milliseconds since input was received on any
// input device.  If unsuccessful, the program prints a message to stderr
// and exits with a non-zero exit code.
//
// Copyright © 2005, 2008 Magnus Henoch           <henoch@dtek.chalmers.se>
// Copyright © 2006, 2007 Danny Kukawka           <danny.kukawka@web.de>
// Copyright © 2008       Eivind Magnus Hvidevold <hvidevold@gmail.com>
// Copyright © 2015       Remy Goldschmidt        <taktoa@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of version 2 of the GNU General Public License
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to
//
//     Free Software Foundation, Inc.,
//     51 Franklin Street, Fifth Floor
//     Boston, MA 02110-1301, USA.
//
// The function workaroundX11 was adapted from kpowersave-0.7.3 by
// Eivind Magnus Hvidevold <hvidevold@gmail.com>.
// kpowersave is licensed under the GNU GPL, version 2 _only_.

#include <X11/Xlib.h>
#include <X11/extensions/dpms.h>
#include <X11/extensions/scrnsaver.h>
#include <stdbool.h>
#include <signal.h>
#include <time.h>
#include <config.h>
#include "system.h"
#include "errno.h"
#include "error.h"
#include "progname.h"
#include "xalloc.h"
#include "time-format.h"

// Macros

#define AUTHORS          "Magnus Henoch"

#define TIME_FMT_DEFAULT "%tms"


// Forward declarations.
static ulong workaroundX11(Display *dpy, ulong _idleTime);
static void  check_xss_supported(int event_basep, int error_basep);
static void  try_xss_query(XScreenSaverInfo* ssi);
static void  print_idle_time();
static void  signal_callback_handler(int sig,
                                     siginfo_t *siginfo,
                                     void *context);
static void  fmbprint(FILE* stream, const char* msg);
static void  print_help(FILE *out);
static void  print_version(FILE *out);

const char* time_format;
Display *dpy;

// Open the display
static void open_display() {
    dpy = XOpenDisplay(NULL);
    if(dpy == NULL) {
        fprintf(stderr, _("Error: couldn't open display\n"));
        exit(EXIT_FAILURE);
    }
}

// Check if the XScreenSaver extension is supported by the X server
static void check_xss_supported(int event_basep, int error_basep) {
    if(!XScreenSaverQueryExtension(dpy, &event_basep, &error_basep)) {
        fprintf(stderr, _("Error: screen saver extension not supported\n"));
        exit(EXIT_FAILURE);
    }
}

// Try to query the current idle time
static void try_xss_query(XScreenSaverInfo* ssi) {
    if(!XScreenSaverQueryInfo(dpy, DefaultRootWindow(dpy), ssi)) {
        fprintf(stderr, _("Error: couldn't query screen saver info\n"));
        exit(EXIT_FAILURE);
    }
}

// Get the current idle time
static ulong get_idle_time() {
    // Open display
    open_display();

    // Allocate XScreenSaverInfo variable
    XScreenSaverInfo* ssi = XScreenSaverAllocInfo();

    // Pointers for event and error results
    int event_basep, error_basep;

    // Create sigaction struct
    struct sigaction act;
    memset(&act, 0, sizeof(act));

    // Use the sa_sigaction field because the handles
    // have two additional parameters
    act.sa_sigaction = &signal_callback_handler;

    // The SA_SIGINFO flag tells sigaction() to use the
    // sa_sigaction field, not sa_handler.
    act.sa_flags = SA_SIGINFO;

    // Register signal and signal handler
    if(sigaction(SIGTERM, &act, NULL) < 0) {
        perror(_("Error: sigaction"));
        exit(EXIT_FAILURE);
    }

    // Enable line buffering on stdout
    setlinebuf(stdout);

    // Get the current idle time
    check_xss_supported(event_basep, error_basep);
    try_xss_query(ssi);

    // Save idle time
    ulong idle = ssi->idle;

    // Free ssi
    XFree(ssi);

    return idle;
}

// Print the current idle time
static void print_idle_time() {
    fmbprint(stdout, timefmt(time_format, get_idle_time()));
}

// Print a string in multibyte format
static void fmbprint(FILE* stream, const char* msg) {
    size_t len = mbsrtowcs(NULL, &msg, 0, NULL);
    if(len == (size_t)-1) {
        error(EXIT_FAILURE, errno, _("conversion to multibyte string failed"));
    }
    wchar_t* mb_msg = xmalloc((len + 1) * sizeof(wchar_t));
    mbsrtowcs(mb_msg, &msg, len + 1, NULL);
    fwprintf(stream, L"%ls\n", mb_msg);
    free(mb_msg);
}

// Signal callback handler
static void signal_callback_handler(int sig,
                                    siginfo_t* siginfo,
                                    void* context) {
    XCloseDisplay(dpy);
}

// Print help info. The message is split into pieces for the translators.
static void print_help(FILE* out) {
    const char *lc_messages = setlocale(LC_MESSAGES, NULL);
    /* TRANSLATORS: --help output 1 (synopsis)
       no-wrap */
    fprintf(out, _("Usage: %s [OPTION]...\n"), program_name);
    /* TRANSLATORS: --help output 2 (brief description)
       no-wrap */
    fputs(_("Prints the user's X11 idle time.\n"), out);
    fputs("\n", out);
    /* TRANSLATORS: --help output 3: options
       no-wrap */
    fputs(_(" -f, --format[=]FMT  use FMT as the time format\n"), out);
    fputs(_(" -h, --help          display this help\n"), out);
    fputs(_("     --version       output version information\n"), out);
    fputs("\n", out);
    /* TRANSLATORS: --help output 4+ (reports)
       TRANSLATORS: the placeholder indicates the bug-reporting address
       for this application.
       no-wrap */
    fprintf(out, _("Report bugs to: %s\n"),            BUG_URL);
    fprintf(out, _("Report translation bugs to %s\n"), TRANSLATION_URL);
    fprintf(out, _("xprintidle-ng home page: %s\n"),   PROJECT_URL);
    exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

// Print version and copyright information.
static void print_version(FILE *out) {
    fprintf(out, "%s (%s) %s\n", PACKAGE, PACKAGE_NAME, VERSION);
    /* xgettext: no-wrap */
    fputs("", out);

    // It is important to separate the year from the rest of the message,
    // as done here, to avoid having to retranslate the message when a new
    // year comes around.
    fprintf(out, _("Copyright © %d %s"), COPYRIGHT_YEAR, AUTHORS);
    fputs(_("License GPLv2: GNU GPL version 2\n"), out);
    fputs(_("<http://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html>"), out);
    fputs(_("This is free software: "
            "you are free to change and redistribute it.\n"), out);
    fputs(_("There is NO WARRANTY, to the extent permitted by law.\n"), out);
}


int main(int argc, char* argv[]) {
    int lose = 0;

    // Option enum
    enum {
        OPT_HELP = CHAR_MAX + 1,
        OPT_VERSION
    };

    // Option format struct
    static const struct option longopts[] = {
        {"format",  required_argument, NULL, 'f'},
        {"help",    no_argument,       NULL, OPT_HELP},
        {"version", no_argument,       NULL, OPT_VERSION},
        {NULL,      0,                 NULL, 0}
    };

    // Set program name to $0
    set_program_name(argv[0]);

    // Set locale via LC_ALL.
    setlocale(LC_ALL, "");

    if(ENABLE_NLS) {
        // Set the text message domain.
        bindtextdomain(PACKAGE, LOCALEDIR);
        textdomain(PACKAGE);
    }

    // Close stdout upon exiting
    atexit(close_stdout);

    char* time_format_raw = _(TIME_FMT_DEFAULT);

    int optc;
    while((optc = getopt_long(argc, argv, "f:", longopts, NULL)) != -1) {
        switch(optc) {
        case OPT_VERSION:
            print_version(stdout);
            exit(EXIT_SUCCESS);
            break;
        case OPT_HELP:
            print_help(stdout);
            exit(EXIT_SUCCESS);
            break;
        case 'f':
            time_format_raw = optarg;
            break;
        default:
            lose = true;
            break;
        }
    }

    if(lose || optind < argc) {
        // Print error message and exit.
        error(0, 0, "%s: %s", _("extra operand"), argv[optind]);
        print_help(stderr);
    }

    time_format = time_format_raw;

    //test_trie();

    print_idle_time();

    exit(EXIT_SUCCESS);
}


src/xprintidle.c
// This program prints the "idle time" of the user to stdout.  The "idle
// time" is the number of milliseconds since input was received on any
// input device.  If unsuccessful, the program prints a message to stderr
// and exits with a non-zero exit code.
//
// Copyright © 2005, 2008 Magnus Henoch           <henoch@dtek.chalmers.se>
// Copyright © 2006, 2007 Danny Kukawka           <danny.kukawka@web.de>
// Copyright © 2008       Eivind Magnus Hvidevold <hvidevold@gmail.com>
// Copyright © 2015       Remy Goldschmidt        <taktoa@gmail.com>
//
// This program is free software; you can redistribute it and/or modify
// it under the terms of version 2 of the GNU General Public License
// as published by the Free Software Foundation.
//
// This program is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License
// along with this program; if not, write to
//
//     Free Software Foundation, Inc.,
//     51 Franklin Street, Fifth Floor
//     Boston, MA 02110-1301, USA.
//
// The function workaroundX11 was adapted from kpowersave-0.7.3 by
// Eivind Magnus Hvidevold <hvidevold@gmail.com>.
// kpowersave is licensed under the GNU GPL, version 2 _only_.

#include <X11/Xlib.h>
#include <X11/extensions/dpms.h>
#include <X11/extensions/scrnsaver.h>
#include <stdbool.h>
#include <signal.h>
#include <time.h>
#include <config.h>
#include "system.h"
#include "errno.h"
#include "error.h"
#include "progname.h"
#include "xalloc.h"

#define AUTHORS             "Magnus Henoch"
#define PROJECT_URL         "https://github.com/taktoa/xprintidle-ng"
#define TRANSLATION_URL     "https://github.com/taktoa/xprintidle-ng/issues"
#define BUG_URL             "https://github.com/taktoa/xprintidle-ng/issues"

#define TIME_FORMAT_DEFAULT "%m"

// Forward declarations.
static unsigned long workaroundX11(Display *dpy, unsigned long _idleTime);
static void check_xss_supported(int event_basep, int error_basep);
static void try_xss_query(XScreenSaverInfo ssi);
static void print_idle_time();
static void signal_callback_handler(int sig, siginfo_t *siginfo, void *context);
static void mbprint(const char* msg);
static void print_help(FILE *out);
static void print_version(FILE *out);
const char* time_format;
Display *dpy;

int main(int argc, char* argv[]) {
    int lose = 0;

    // Option enum
    enum {
        OPT_HELP = CHAR_MAX + 1,
        OPT_VERSION
    };

    // Option format struct
    static const struct option longopts[] = {
        {"format",  required_argument, NULL, 'f'},
        {"help",    no_argument,       NULL, OPT_HELP},
        {"version", no_argument,       NULL, OPT_VERSION},
        {NULL, 0, NULL, 0}
    };

    // Set program name to $0
    set_program_name(argv[0]);

    // Set locale via LC_ALL.
    setlocale(LC_ALL, "");

#if ENABLE_NLS
    // Set the text message domain.
    bindtextdomain(PACKAGE, LOCALEDIR);
    textdomain(PACKAGE);
#endif

    // Close stdout upon exiting
    atexit(close_stdout);

    char* time_format_raw = _(TIME_FORMAT_DEFAULT);
    
    int optc;
    while((optc = getopt_long(argc, argv, "f:", longopts, NULL)) != -1) {
        switch(optc) {
        case OPT_VERSION:
            print_version(stdout);
            exit(EXIT_SUCCESS);
            break;
        case OPT_HELP:
            print_help(stdout);
            exit(EXIT_SUCCESS);
            break;
        case 'f':
            time_format_raw = optarg;
            break;
        default:
            lose = true;
            break;
        }
    }
    
    if(lose || optind < argc) {
        // Print error message and exit.
        error(0, 0, "%s: %s", _("extra operand"), argv[optind]);
        print_help(stderr);
    }

    time_format = time_format_raw;

    print_idle_time();

    exit(EXIT_SUCCESS);
}

// Open the display
static void open_display() {
    dpy = XOpenDisplay(NULL);
    if(dpy == NULL) {
        fprintf(stderr, _("Error: couldn't open display\n"));
        exit(EXIT_FAILURE);
    }
}

// Check if the XScreenSaver extension is supported by the X server
static void check_xss_supported(int event_basep, int error_basep) {
    if(!XScreenSaverQueryExtension(dpy, &event_basep, &error_basep)) {
        fprintf(stderr, _("Error: screen saver extension not supported\n"));
        exit(EXIT_FAILURE);
    }
}

// Try to query the current idle time
static void try_xss_query(XScreenSaverInfo ssi) {
    if(!XScreenSaverQueryInfo(dpy, DefaultRootWindow(dpy), &ssi)) {
        fprintf(stderr, _("Error: couldn't query screen saver info\n"));
        exit(EXIT_FAILURE);
    }
}

// Print the current idle time
static void print_idle_time() {
    open_display();

    XScreenSaverInfo ssi;
    int event_basep, error_basep;

    // Create sigaction struct
    struct sigaction act;
    memset(&act, '\0', sizeof(act));

    // Use the sa_sigaction field because the handles
    // have two additional parameters
    act.sa_sigaction = &signal_callback_handler;

    // The SA_SIGINFO flag tells sigaction() to use the
    // sa_sigaction field, not sa_handler.
    act.sa_flags = SA_SIGINFO;

    // Register signal and signal handler
    if(sigaction(SIGTERM, &act, NULL) < 0) {
        perror(_("Error: sigaction"));
        exit(EXIT_FAILURE);
    }

    // Enable line buffering on stdout
    setlinebuf(stdout);

    // Get the current idle time
    check_xss_supported(event_basep, error_basep);
    try_xss_query(ssi);
    printf("%lu\n", workaroundX11(dpy, ssi.idle));
}

// Print a string in multibyte format
static void mbprint(const char* msg) {
    size_t len = mbsrtowcs(NULL, &msg, 0, NULL);
    if(len == (size_t)-1) {
        error(EXIT_FAILURE, errno, _("conversion to multibyte string failed"));
    }
    wchar_t* mb_msg = xmalloc((len + 1) * sizeof(wchar_t));
    mbsrtowcs(mb_msg, &msg, len + 1, NULL);
    wprintf(L"%ls\n", mb_msg);
    free(mb_msg);
}

// Signal callback handler
static void signal_callback_handler(int sig,
                                    siginfo_t *siginfo,
                                    void *context) {
    XCloseDisplay(dpy);
}

// Print help info. The message is split into pieces for the translators.
static void print_help(FILE *out) {
    const char *lc_messages = setlocale(LC_MESSAGES, NULL);
    /* TRANSLATORS: --help output 1 (synopsis)
       no-wrap */
    fprintf(out, _("Usage: %s [OPTION]...\n"), program_name);
    /* TRANSLATORS: --help output 2 (brief description)
       no-wrap */
    fputs(_("Prints the user's X11 idle time.\n"), out);
    fputs("\n", out);
    /* TRANSLATORS: --help output 3: options
       no-wrap */
    fputs(_(" -f, --format[=]FMT  use FMT as the time format\n"), out);
    fputs(_(" -h, --help          display this help\n"), out);
    fputs(_("     --version       output version information\n"), out);
    fputs("\n", out);
    /* TRANSLATORS: --help output 4+ (reports)
       TRANSLATORS: the placeholder indicates the bug-reporting address
       for this application.
       no-wrap */
    fprintf(out, _("Report bugs to: %s\n"),            BUG_URL);
    fprintf(out, _("Report translation bugs to %s\n"), TRANSLATION_URL);
    fprintf(out, _("xprintidle home page: %s\n"),      PROJECT_URL);
    exit(out == stderr ? EXIT_FAILURE : EXIT_SUCCESS);
}

// Print version and copyright information.
static void print_version(FILE *out) {
    fprintf(out, "%s (%s) %s\n", PACKAGE, PACKAGE_NAME, VERSION);
    /* xgettext: no-wrap */
    fputs("", out);

    // It is important to separate the year from the rest of the message,
    // as done here, to avoid having to retranslate the message when a new
    // year comes around.
    fprintf(out, _("Copyright © %d %s"), COPYRIGHT_YEAR, AUTHORS);
    fputs(_("License GPLv2: GNU GPL version 2\n"), out);
    fputs(_("<http://www.gnu.org/licenses/old-licenses/gpl-2.0.en.html>"), out);
    fputs(_("This is free software: "
            "you are free to change and redistribute it.\n"), out);
    fputs(_("There is NO WARRANTY, to the extent permitted by law.\n"), out);
}

/*!
 * This function works around an XServer idleTime bug in the
 * XScreenSaverExtension if dpms is running. In this case the current
 * dpms-state time is always subtracted from the current idletime.
 * This means: XScreenSaverInfo->idle is not the time since the last
 * user activity, as descriped in the header file of the extension.
 * This result in SUSE bug # and sf.net bug #. The bug in the XServer itself
 * is reported at https://bugs.freedesktop.org/buglist.cgi?quicksearch=6439.
 *
 * Workaround: Check if if XServer is in a dpms state, check the
 *             current timeout for this state and add this value to
 *             the current idle time and return.
 *
 * \param  _idleTime   an unsigned long value with the current idletime from
 *                     XScreenSaverInfo->idle
 * \return             an unsigned long with the corrected idletime
 */
static unsigned long workaroundX11(Display *dpy, unsigned long _idle) {
    int dummy;
    CARD16 standby, suspend, off;
    CARD16 state;
    BOOL onoff;

    if(DPMSQueryExtension(dpy, &dummy, &dummy)) {
        if(DPMSCapable(dpy)) {
            DPMSGetTimeouts(dpy, &standby, &suspend, &off);
            DPMSInfo(dpy, &state, &onoff);

            if(onoff) {
                switch(state) {
                case DPMSModeStandby:
                    // this check is a little bit paranoid, but just to be sure
                    if(_idle < (unsigned) (standby * 1000)) {
                        _idle += (standby * 1000);
                    }
                    break;
                case DPMSModeSuspend:
                    if(_idle < (unsigned) ((suspend + standby) * 1000)) {
                        _idle += ((suspend + standby) * 1000);
                    }
                    break;
                case DPMSModeOff:
                    if(_idle < (unsigned) ((off + suspend + standby) * 1000)) {
                        _idle += ((off + suspend + standby) * 1000);
                    }
                    break;
                case DPMSModeOn:
                    break;
                default:
                    break;
                }
            }
        }
    }

    return _idle;
}
